### DISCUSSION TITLE
Problems with meshing non-convex domain

### DISCUSSION THREAD

**Discussion by elisk616** [neperfepx/neper#825] [Q&A] [answered]
Hi! I'm new to Neper and ran into a problem when I tried to run the code for the following shape (taken from the website)

`neper -T -n 1000 -morpho gg -domain "cube(1.2,0.1,1.2)" -transform "cut(cylinder(0.3,0,0.3,0,1,0,0.15),cylinder(0.3,0,0.9,0,1,0,0.15),cylinder(0.9,0,0.3,0,1,0,0.15),cylinder(0.9,0,0.9,0,1,0,0.15))" -o 4_holes`
![Screenshot 2024-03-06 131157](https://github.com/neperfepx/neper/assets/160881304/22df2a3c-f8e3-4e21-a28b-2f5c931b13f7)

It generated a .tess-file (although very slowly, it took 6929.048 secs, is that normal?) but I can't mesh it. It aborts during the 2D meshing, can anyone help me with this issue?

`neper -M 4_holes.tess -faset none -format inp`

========================    N   e   p   e   r    =======================
Info   : A software package for polycrystal generation and meshing.
Info   : Version 4.8.1
Info   : Built with: gsl|muparser|opengjk|openmp|nlopt|libscotch (full)
Info   : Running on 48 threads.
Info   : <https://neper.info>
Info   : Copyright (C) 2003-2022, and GNU GPL'd, by Romain Quey.
Info   : No initialization file found (`/home/z004wttr/.neperrc').

Info   : ---------------------------------------------------------------
Info   : MODULE  -M loaded with arguments:
Info   : [ini file] (none)
Info   : [com line] 4_holes.tess -faset none -format inp
Info   : ---------------------------------------------------------------
Info   : Reading input data...
Info   :   - Reading arguments...
Info   : Loading input data...
Info   :   - Loading tessellation...
Info   :     [i] Parsing file `4_holes.tess'...
Info   :     [i] Parsed file `4_holes.tess'.
Info   : Meshing...
Info   :   - Preparing... (cl = 0.02654) 100%
Info   :   - 0D meshing... 100%
Info   :   - 1D meshing... 100%
Info   :   - 2D meshing...  29% (0.25|0.83/95%| 4%| 1%) 
*pbnodeqty = 11
nodeqty1d = 14
min_dist_max = 0.000000
Aborted (core dumped)

**Comment by rquey** [comment, marked as answer]
It takes 2500 secs on my laptop.

These tessellations do not *necessarily* mesh properly. You first need to apply regularization (`-reg 1`), but even so, there may be problems.  If meshing fails, try to change the tessellation (different `-id`).
